pkg/dns/dns.go

```diff
@@ -282,6 +282,21 @@ func (h *Handler) ServeDNS(w dns.ResponseWriter, req *dns.Msg) {
 		h.handleDefault(w, req)
 	}
 }
+func StartWithListener(opts ServerOptions) (*Server, error) {
+	h, err := newHandler(opts.IPv6, opts.StaticHosts, opts.UpstreamServers)
+	if err != nil {
+		return nil, err
+	}
+	server := &Server{}
+	s := &dns.Server{Net: "tcp", Listener: opts.Listener, Handler: h}
+	server.tcp = s
```

Comment:
Why do we need the extra assignment?
```suggestion
	s := &dns.Server{Net: "tcp", Listener: opts.Listener, Handler: h}
	server := &Server{tcp: s}
```